Sources/LanguageServerProtocol/CodeAction.swift

```diff
@@ -0,0 +1,102 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2018 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+/// Request for returning all possible code actions for a given text document and range.
+///
+/// The code action request is sent from the client to the server to compute commands for a given text
+/// document and range. These commands are typically code fixes to either fix problems or to beautify/
+/// refactor code.
+///
+/// Servers that provide code actions should set the `codeActions` server capability.
+///
+/// - Parameters:
+///   - textDocument: The document in which the command was invoked.
+///   - range: The specific range inside the document to search for code actions.
+///   - context: The context of the request.
+///
+/// - Returns: A list of folding ranges for the given document.
+public struct CodeActionRequest: TextDocumentRequest, Hashable {
+  public static let method: String = "textDocument/codeAction"
+  public typealias Response = [CodeAction]?
+
+  /// The range for which the command was invoked.
+  public var range: PositionRange
+
+  /// Context carrying additional information.
+  public var context: CodeActionContext
+
+  /// The document in which the command was invoked.
+  public var textDocument: TextDocumentIdentifier
+
+  public init(range: PositionRange, context: CodeActionContext, textDocument: TextDocumentIdentifier) {
```

Comment:
I'm trying to minimize use of `PositionRange`, so I suggest making the initializer take `Range<Position>` and convert it.  The only reason we have PositionRange is to make the serialization work.  Ideally it would just be an implementation detail, but for now it's part of the API so that we can keep getting `Codable` synthesis.